File: love2d/src/game/levelHandler.lua
levelHandler = {}

function levelHandler:load()
    self.currentLevel = 0
    self.manSpawnX = 0
    self.manSpawnY = 0
    self.currentTime = 0
    self:init()
end

function levelHandler:init()
    self:clean()
    self.level = sti("maps/level_"..self.currentLevel..".lua")
    self.objects = self.level.layers["Objects"]
    self.groundLayer = self.level.layers["Ground"]
    self.solidObjects = self.level.layers["Solid"]
    
    for i,v in pairs(self.solidObjects.objects) do
        local solid = fcd:newItem("Solid",v.x,v.y,v.width,v.height,"Solid")
    end
    for i,v in pairs(self.objects.objects) do
        self:setObject(v)
    end
end

function levelHandler:draw(mul)
    self.level:draw(0,0,mul,mul)
    shard.drawAll()
    roamer.drawAll()
    spike.drawAll()
    jones.drawAll()
    beans.drawAll()
    glassdoor.drawAll()
    flag:draw()
end

function levelHandler:update(dt)
    if self.currentLevel ~= 0 and self.currentLevel ~= 18 then
        self.currentTime = self.currentTime+dt
        
    end
    roamer.updateAll(dt)
    shard.updateAll(dt)
    jones.updateAll()
    beans.updateAll()
end

function levelHandler:setObject(obj)
    if obj.type == "man" then
        self.manSpawnX = obj.x+4
        self.manSpawnY = obj.y+4
        man:respawn(self.manSpawnX, self.manSpawnY)
    elseif obj.type == "flag" then
        flag:move(obj.x,obj.y)
    elseif obj.type == "spike" then
        spike.new(obj.x,obj.y)
    elseif obj.type == "shard" then
        shard.new(obj.x,obj.y)
    elseif obj.type == "ghost" then
        roamer.new(obj.x,obj.y)
    elseif obj.type == "jones" then
        jones.new(obj.x,obj.y)
    elseif obj.type == "beans" then
        beans.new(obj.x+2,obj.y)
    elseif obj.type == "glassdoor" then
        glassdoor.new(obj.x,obj.y)
    end
    if self.currentLevel >= 17 then
        flag:move(1024,1024)
    end
    if self.currentLevel == 18 then
        love.audio.play(audio.bgm.End)
        text:make(26,46,"congratulations\\time:"..math.floor(self.currentTime))
    end
end
function levelHandler:clean()
    fcd.items.Solid = {}
    spike.destroyAll()
    shard.destroyAll()
    roamer.destroyAll()
    corpse.removeAll()
    jones.destroyAll()
    beans.destroyAll()
    glassdoor.destroyAll()

end
function levelHandler:reset()
    self:goTo(self.currentLevel)
end

function levelHandler:goTo(level)
    self.currentLevel = level
    self:init()
end

File: windows/LevelHandler.cpp
#include "LevelHandler.hpp"
LevelHandler levelhandler;

LevelHandler::LevelHandler() {

}

Vector2 LevelHandler::GetSpawnLocation() {
	return PlayerSpawnLocation;
}

void LevelHandler::Start() {
	CurrentLevel = 0;
	Level = "res/maps/level_" + std::to_string(CurrentLevel) + ".json";
	TiledLevel = LoadTiled(Level.c_str());
	CurrentTime = 0;
	text.Init();
	Init();
}

LevelHandler::~LevelHandler() {
	
}

void LevelHandler::Init() {
	Clean();
	UnloadMap(TiledLevel);
	Level = "res/maps/level_" + std::to_string(CurrentLevel) + ".json";
	TiledLevel = LoadTiled(Level.c_str());
	std::ifstream LevelFile(Level.c_str());
	if (CurrentLevel == 18) {
		text.Make("CONGRATULATIONS@TIME:" + std::to_string(int(CurrentTime)));
	}
	if (!LevelFile.is_open()) {
		std::cerr << "Nah mate\n";
		
	}
	else {
		std::cout << "Ye\n";
	}
	LevelFile >> LevelData;
	for (const auto& item : LevelData["layers"]) {
		if (item["type"] == "objectgroup") {
			if (item["name"] == "Solid") {
				for (const auto& object : item["objects"]) {
					CollisionDetection.newItem("Solid", (float)object["x"], (float)object["y"], (float)object["width"], (float)object["height"], Solid);
				}
			}
		}
	}
	for (const auto& item : LevelData["layers"]) {
		if (item["type"] == "objectgroup") {
			if (item["name"] == "Objects") {
				for (const auto& object : item["objects"]) {
					SetObject(object);
				}
			}
		}
	}
}

void LevelHandler::SetObject(auto obj) {
	if (obj["type"] == "man") {
		PlayerSpawnLocation.x = (float)obj["x"] + 4.0f;
		PlayerSpawnLocation.y = (float)obj["y"] + 4.0f;
		man.respawn(PlayerSpawnLocation.x, PlayerSpawnLocation.y);
	}
	else if (obj["type"] == "flag")
		objects.push_back(std::make_unique<Flag>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "spike")
		enemies.push_back(std::make_unique<Spike>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "shard")
		enemies.push_back(std::make_unique<Shard>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "ghost")
		enemies.push_back(std::make_unique<Roamer>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "jones")
		objects.push_back(std::make_unique<Jones>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "beans")
		objects.push_back(std::make_unique<Beans>(Vector2{ (float)obj["x"], (float)obj["y"] }));
	else if (obj["type"] == "glassdoor")
		objects.push_back(std::make_unique<StoreDoor>(Vector2{ (float)obj["x"], (float)obj["y"] }));
}

void LevelHandler::Clean() {
	if (!CollisionDetection.items.empty()) {
		CollisionDetection.deleteClass(Solid);
	}
	enemies.clear();
	objects.clear();
}

void LevelHandler::SpawnCorpse(Vector2 pos, Vector2 velocity) {
	objects.push_back(std::make_unique<Corpse>(pos,velocity));
}

int LevelHandler::GetLevel() {
	return CurrentLevel;
}

void LevelHandler::SetLevel(int level) {
	CurrentLevel = level;
	Init();
}
void LevelHandler::NextLevel() {
	CurrentLevel++;
	Init();
}
void LevelHandler::PrevLevel() {
	CurrentLevel--;
	Init();
}
void LevelHandler::DrawLevel() {
	if (!Level.empty() && IsTiledReady(TiledLevel)) {
		DrawTiled(TiledLevel, 0, 0, WHITE);
	}
	for (auto& enemy : enemies) {
		enemy->Draw();
	}
	for (auto& object : objects) {
		object->Draw();
	}
	text.Draw();
}
void LevelHandler::Update() {
	CurrentTime += GetFrameTime();
	for (auto& enemy : enemies) {
		enemy->Update();
	}
	for (auto& object : objects) {
		object->Update();
	}
	text.Update();
}